Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstructionModel.py

```diff
@@ -0,0 +1,165 @@
+﻿# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+AddReference("QuantConnect.Indicators")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Indicators import *
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from QuantConnect.Algorithm.Framework.Portfolio import PortfolioTarget
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+
+### <summary>
+### Provides an implementation of Mean-Variance portfolio optimization based on modern portfolio theory.
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the last three months daily price to calculate the optimal weight 
+### with the weight range from -1 to 1 and minimize the portfolio variance with a target return of 2%
+### </summary>
+class MeanVarianceOptimizationPortfolioConstructionModel:
+    def __init__(self, *args, **kwargs):
+        """ 
+        Initialize the model
+        Args:
+            resolution: The resolution of the history price 
+            period(int): The time interval of history price to calculate the weight 
+            minimum_weight(float): The lower bounds on portfolio weights
+            maximum_weight(float): The upper bounds on portfolio weights
+            target_return(float): The target portfolio return
+            optimization_method(method): Method used to compute the portfolio weights
+        """
+        self.resolution = kwargs['resolution'] if 'resolution' in kwargs else Resolution.Daily
+        self.period = kwargs['period'] if 'period' in kwargs else 63
+        self.minimum_weight = kwargs['minimum_weight'] if 'minimum_weight' in kwargs else -1
+        self.maximum_weight = kwargs['maximum_weight'] if 'maximum_weight' in kwargs else 1
+        self.target_return = kwargs['target_return'] if 'target_return' is kwargs else 2
```

Comment:
I think this is daily?

Reply:
We use annual returns to compute the weights and consequently this is annual target return.

Reply:
Should we use percentage return 0.02?